File: rust/2015/src/day24.rs
use itertools::Itertools;

fn parse(filename: &str) -> Vec<u32> {
    let mut packages = aoc::file::read_numbers::<u32>(filename);
    packages.reverse();
    packages
}

type Group = Vec<u32>;

fn qe(group: &Group) -> u64 {
    group.iter().map(|w| *w as u64).product::<u64>()
}

fn solve_case<const NUM_GROUPS: u32>(packages: &[u32]) -> u64 {
    // Assume all packages have unique weights and are sorted in decreasing order
    let total_weight = packages.iter().sum::<u32>();
    debug_assert!(
        (total_weight % NUM_GROUPS) == 0,
        "Invalid package configuration"
    );
    let target_group_weight = total_weight / NUM_GROUPS;
    let mut num_packages = 0;
    let mut weight = 0;
    let size = packages.len();
    while (num_packages < size) {
        weight += packages[num_packages];
        if (weight > target_group_weight) {
            break;
        }
        num_packages += 1;
    }
    let mut possible = Vec::<Group>::new();
    for index in num_packages - 1..size {
        let neighbors = packages
            .iter()
            .combinations(index + 1)
            .filter(|combo| combo.iter().map(|w| **w).sum::<u32>() == target_group_weight)
            .map(|combo| {
                let mut combo = combo.iter().map(|w| **w).collect::<Vec<_>>();
                combo.sort();
                combo
            })
            .collect::<Vec<_>>();
        if (!neighbors.is_empty()) {
            possible.extend(neighbors);
            break;
        }
    }
    possible.sort_by(|lhs, rhs| lhs.len().cmp(&rhs.len()).then(qe(lhs).cmp(&qe(rhs))));
    return qe(&possible[0]);
}

fn main() {
    println!("Part 1");
    let example = parse("day24.example");
    assert_eq!(99, solve_case::<3>(&example));
    let input = parse("day24.input");
    assert_eq!(10723906903, solve_case::<3>(&input));

    println!("Part 2");
    assert_eq!(44, solve_case::<4>(&example));
    assert_eq!(74850409, solve_case::<4>(&input));
}

File: 2015/day24.cpp
// https://adventofcode.com/2015/day/24

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <print>
#include <ranges>
#include <span>

auto parse(String const& filename) -> Vec<u32> {
  auto packages =
      aoc::views::read_numbers<u32>(filename) | aoc::collect_vec<u32>();
  stdr::reverse(packages);
  return packages;
}

// TODO: aoc::static_vector<u32, 16>
using Group = Vec<u32>;

fn qe(std::span<const u32> group) -> u64 {
  return stdr::fold_left(group, u64{1}, std::multiplies{});
}

template <u32 NUM_GROUPS>
fn solve_case(std::span<const u32> packages) -> u64 {
  // Assume all packages have unique weights and are sorted in decreasing order
  let total_weight = aoc::ranges::accumulate(packages, 0u);
  AOC_ASSERT((total_weight % NUM_GROUPS) == 0, "Invalid package configuration");
  let target_group_weight = total_weight / NUM_GROUPS;
  auto num_packages = 0uz;
  auto weight = 0u;
  let size = packages.size();
  while (num_packages < size) {
    weight += packages[num_packages];
    if (weight > target_group_weight) {
      break;
    }
    num_packages += 1;
  }
  auto possible = Vec<Group>{};
  for (let index : Range{num_packages - 1, size}) {
    possible = packages |
               aoc::views::combinations(index + 1) |
               stdv::filter([&](let& combo) {
                 return aoc::ranges::dot_product(packages, combo) ==
                        target_group_weight;
               }) |
               stdv::transform([&](let& combo) {
                 let neighbor = aoc::binary_select_from_combination<Group>(
                     packages, combo);
                 return neighbor;
               }) |
               aoc::collect_vec<Group>();
    if (!possible.empty()) {
      break;
    }
  }
  stdr::sort(possible, [](Group const& lhs, Group const& rhs) {
    if (lhs.size() == rhs.size()) {
      return qe(lhs) < qe(rhs);
    }
    return lhs.size() < rhs.size();
  });
  return qe(possible[0]);
}

int main() {
  std::println("Part 1");
  let example = parse("day24.example");
  AOC_EXPECT_RESULT(99, solve_case<3>(example));
  let input = parse("day24.input");
  AOC_EXPECT_RESULT(10723906903, solve_case<3>(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(44, solve_case<4>(example));
  AOC_EXPECT_RESULT(74850409, solve_case<4>(input));

  AOC_RETURN_CHECK_RESULT();
}
